app/lib/analyzer/backend.dart

```diff
@@ -110,33 +111,52 @@ class AnalysisBackend {
           .append(Package, id: packageName)
           .append(PackageVersion, id: packageVersion)
     ]);
-    // package or version does not exists
-    if (versions.first == null) return false;
+    // Does package and version exist?
+    final PackageVersion pv = versions.single;
+    if (pv == null) return false;
 
+    // Does package have any analysis?
     final List<PackageAnalysis> list =
         await db.lookup([db.emptyKey.append(PackageAnalysis, id: packageName)]);
-    // no analysis for the package
-    if (list.first == null) return true;
-
     final PackageAnalysis packageAnalysis = list.first;
+    if (packageAnalysis == null) return true;
+
+    // Does package have newer version than latest analyzed version?
+    if (isNewer(packageAnalysis.latestSemanticVersion, pv.semanticVersion)) {
+      return true;
+    }
+
+    // Does package have analysis for the current version?
     final PackageVersionAnalysis versionAnalysis =
         (await db.lookup([packageAnalysis.latestVersionKey])).first;
+    if (versionAnalysis == null) return true;
 
-    if (versionAnalysis.analysisVersion == analysisVersion) {
-      final DateTime now = new DateTime.now().toUtc();
-      if (now.difference(versionAnalysis.analysisTimestamp).inHours < 4) {
-        // latest analysis has the same version and happened in the past 4 hours
-        return false;
-      }
-    } else {
-      final Version analysisSemantic = new Version.parse(analysisVersion);
-      if (isNewer(analysisSemantic, versionAnalysis.semanticAnalysisVersion)) {
-        // existing analysis version is newer, current analyzer is obsolete?
-        // TODO: turn off task polling on this instance
-        _logger.warning(
-            'Newer analysis version detected, current instance is obsolete.');
-        return false;
-      }
+    // Is current analysis version newer?
+    final Version analysisSemantic = new Version.parse(analysisVersion);
+    if (isNewer(versionAnalysis.semanticAnalysisVersion, analysisSemantic)) {
+      return true;
+    }
+
+    // Is current analysis version obsolete?
+    if (isNewer(analysisSemantic, versionAnalysis.semanticAnalysisVersion)) {
+      // existing analysis version is newer, current analyzer is obsolete?
+      // TODO: turn off task polling on this instance
+      _logger.warning(
+          'Newer analysis version detected, current instance is obsolete.');
+      return false;
+    }
+
+    assert(versionAnalysis.analysisVersion == analysisVersion);
+
+    // Was latest analysis aborted?
+    if (versionAnalysis.analysisStatus == AnalysisStatus.aborted) {
+      return true;
+    }
+
+    // Is latest analysis older than 4 hours?
+    final DateTime now = new DateTime.now().toUtc();
+    if (now.difference(versionAnalysis.analysisTimestamp).inHours < 4) {
```

Comment:
That is true, we can restrict that. But even then, we could have duplicates e.g. when
- (a) first the trigger comes in via HTTP
- (b) the analyzer starts working on it, but there is no sign of it in the DB yet
- (c) the head poller source creates a task too, because it notices the new version
- (d) the analyzer completes the task, and schedules the next one, which happens to be the same package and version.